### PR TITLE
Speculator training code refresh

### PR DESCRIPTION
**Pull request by sahilsuneja1**
now supporting TP base models as well

### CODE REVIEW

File: fms_fsdp/config/training.py [hunk 1]
```diff
@@ -59,3 +59,13 @@ class train_config:
 
     # compile
     use_torch_compile: bool = True
+
```
Comment (JRosenkranz) [inline fms_fsdp/config/training.py:62]: Are we re-using all/most of the above for Speculator training or would it make more sense to have a separate speculator training configuration?

File: fms_fsdp/utils/checkpointing_utils.py [hunk 2]
```diff
@@ -168,7 +172,14 @@ def _validate_ckp_path(self, path):
...
     def load(
-        self, model, optimizer, dataloader, path="", reset_stepcount=False, strict=True
+        self,
+        model,
+        optimizer,
+        dataloader,
+        path="",
+        reset_stepcount=False,
+        strict=True,
+        **kwargs,
```
Comment (JRosenkranz) [inline fms_fsdp/utils/checkpointing_utils.py:182]: might want to just add an `is_compiled=False` parameter to this load. I think it complicates it being a kwarg

File: fms_fsdp/utils/checkpointing_utils.py [hunk 3]
```diff
@@ -193,8 +204,18 @@ def load(
             model_load_time = time.time()
             if os.path.isfile(load_path):
                 checkpoint_data = torch.load(load_path, map_location="cpu")
-                model.load_state_dict(checkpoint_data.get("model_state"), strict=strict)
-                model.to(self.local_rank)
+                if "is_compiled" in kwargs.keys() and kwargs["is_compiled"] is True:
```
Comment (JRosenkranz) [inline fms_fsdp/utils/checkpointing_utils.py:207]: if the above is an argument, this conditional can be made simpler

File: fms_fsdp/utils/checkpointing_utils.py [hunk 4]
```diff
@@ -296,7 +320,10 @@ def save_single_file(
             StateDictType.FULL_STATE_DICT,
             FullStateDictConfig(offload_to_cpu=True, rank0_only=True),
         ):
-            model_state = model.state_dict()
+            if "is_compiled" in kwargs.keys() and kwargs["is_compiled"] is True:
```
Comment (JRosenkranz) [inline fms_fsdp/utils/checkpointing_utils.py:323]: if the above is an argument, this conditional can be made simpler

File: fms_fsdp/utils/dataloader_utils.py [hunk 5]
```diff
@@ -19,6 +19,18 @@
 }
 
 
+def causal_lm(data_seq, prompt_len=1):
+    """
+    Perform causal language modeling by right-shifting the input sequence.
+    Sets first prompt_len tokens to be ignored by the loss.
+    """
+    data_seq = torch.IntTensor(data_seq)
```
Comment (JRosenkranz) [inline fms_fsdp/utils/dataloader_utils.py:27]: do we need to add device to this?

File: fms_fsdp/utils/dataloader_utils.py [hunk 6]
```diff
@@ -56,21 +68,13 @@ def get_data_loader(cfg, rank, world_size):
         Rank of current distributed worker. Used for handling dataset sharding logic.
     world_size : int
         Number of distributed workers. Used for handling dataset sharding logic.
+    postprocess : List[Callable]
```
Comment (JRosenkranz) [inline fms_fsdp/utils/dataloader_utils.py:71]: can we include input/output types here

File: scripts/train_speculator.sh [hunk 7]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+
+# On AWS, the EFA and OFI paths enable NCCL to use optimized networking.
+export LD_LIBRARY_PATH=/opt/nccl/build/lib:/opt/amazon/efa/lib:/opt/amazon/openmpi/lib:/opt/aws-ofi-nccl/lib:/usr/local/cuda/lib:/usr/local/cuda/lib64:/usr/local/cuda:/usr/local/cuda/targets/x86_64-linux/lib/:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/lib:$LD_LIBRARY_PATH
+
+export FI_EFA_SET_CUDA_SYNC_MEMOPS=0
+
+MODEL_ARGS="\
```
Comment (JRosenkranz) [inline scripts/train_speculator.sh:8]: We may want to have a readme for the speculator parameters specifically so a user knows what to explicitly change in this script

File: speculator/train_speculator.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+    setup_environ_flags,
+)
+from speculator.train_speculator_utils import train_speculator
+
+
+os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
+
+
+def test_model(rank, model, arch, cfg, prompt_type="chat"):
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:34]: at some point it would be good to have an assertion added to this such that the training will not start without this passing

File: speculator/train_speculator.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+    torch.set_default_dtype(torch.bfloat16)
+
+    # get policy
+    (
+        mixed_precision_policy,
+        wrapping_policy,
+        sharding_strategy_policy,
+        apply_selective_ac,
+        param_init_fn,
+    ) = get_policies(cfg, rank, LLaMABlock)
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:115]: This is currently hardcoded to LLaMABlock, should it be based on the architecture?

File: speculator/train_speculator.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+
+    model = get_model(
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:117]: Currently we are assuming source="hf", might make sense to try the new `get_model` from https://github.com/foundation-model-stack/foundation-model-stack/pull/326

File: speculator/train_speculator.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+    )
+
+    if rank == 0:
+        print(f"{time.time()}")
+        print(model.config)
+        print(model)
+
+    model.eval()
+    with torch.no_grad():
+        test_model(rank, model, cfg.model_arch, cfg, prompt_type="chat")
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:136]: can we make prompt_type configurable through the speculator training config?

File: speculator/train_speculator.py [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+
+    if hasattr(model.config, "emb_dim"):
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:138]: pieces like extracting emb_dim and vocab_size could probably be moved out to helper functions to make this method shorter for readability

File: speculator/train_speculator.py [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+
+    # get speculator
+    if rank == 0:
+        print(f"{time.time()} Loading speculator")
+    speculator = MLPSpeculator(
+        emb_dim,
+        cfg.speculator_width,
+        vocab_size,
+        cfg.n_speculator_heads,
+        tie_weights=True,
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:162]: should `tie_weights` and `scale_input` be configurable?

File: speculator/train_speculator.py [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+    if rank == 0:
+        total_params = sum(
+            p.numel() for p in speculator.parameters() if p.requires_grad
+        )
+        print(f"\n{time.time()} speculator has {total_params / 1e6} Million params\n")
+
+    # get data loader
+    if rank == 0:
+        print(f"{time.time()} Constructing datasets...")
+    if not cfg.use_dummy_dataset:
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:176]: same here might want to move out to helper function

File: speculator/train_speculator.py [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+        is_compiled=cfg.use_torch_compile,
+    )
+
+    # LR schedule
+    # These functions provide LR scaling factors in [0,1] based on step count.
+    # Stage 1: warm up over first 2k or 5% of steps, whichever is smaller.
+    # Then cosine anneal to 10% of max LR.
+    warmup_interval1 = min(2000, cfg.stage2_start_step // 20)
+    stage1_schedule = lambda x: min(
+        1 - (1 - min(x, warmup_interval1) / warmup_interval1) ** 2,
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:259]: can this be explained a bit more

File: speculator/train_speculator.py [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+            1
+            + math.cos(min(x, cfg.stage2_start_step) / cfg.stage2_start_step * math.pi)
+        ),
+    )
+    # Stage 2: warm up over first 2k or 5% of steps, whichever is smaller.
+    # Then cosine anneal to 10% of stage 1's final LR.
+    warmup_interval2 = min(2000, (cfg.num_steps - cfg.stage2_start_step) // 20)
+    stage2_schedule = lambda x: min(
+        0.1 * (1 - (1 - min(x, warmup_interval2) / warmup_interval2) ** 2),
+        0.01
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:273]: can this be explained a bit more as well

File: speculator/train_speculator_utils.py [hunk 17]
```diff
@@ -0,0 +1,546 @@
...
+    scheduler: torch.optim.lr_scheduler.LRScheduler
+        A scheduler for the optimizer's LR. Scheduler.step() is called on every optimizer step.
+    checkpointer: fms_fsdp.utils.checkpointing_utils.Checkpointer
+        A checkpointer tied to the save directory. Used for saving distributed checkpoints.
+    start_step: optional[int]
+        If resuming from checkpoint, resume step count from this value.
+    n_tok: optional[int]
+        If resuming from checkpoint, resume token count from this value.
+    profiler: optional[torch.profiler.profile]
+        Optional torch profiler for performance benchmarking.
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:287]: base_model_mesh missing from docs

File: speculator/train_speculator_utils.py [hunk 18]
```diff
@@ -0,0 +1,546 @@
...
+        optimizer.step()
+        scheduler.step()
+
+        ddp_stats[1] += 1
+
+        if profiler:
+            profiler.step()
+
+        if batch_idx % cfg.report_interval == 0:
+            dist.all_reduce(ddp_stats, op=dist.ReduceOp.SUM)
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:353]: may want to move this out into a helper function

File: speculator/train_speculator_utils.py [hunk 19]
```diff
@@ -0,0 +1,546 @@
...
+from fms.utils.generation import _make_cache_contiguous
+from torch.nn import CrossEntropyLoss
+from torch.utils.data import DataLoader
+
+from fms_fsdp.config import train_config
+from fms_fsdp.utils.checkpointing_utils import Checkpointer
+from fms_fsdp.utils.config_utils import get_model_config
+
+
+def generate(
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:28]: In `fms` we added a post processing hook to generate, we may be able to re-use that and remove this method entirely. https://github.com/foundation-model-stack/foundation-model-stack/pull/324

File: speculator/train_speculator_utils.py [hunk 20]
```diff
@@ -0,0 +1,546 @@
...
+    if not batched:
+        result = result[0]
+
+    if include_embeds:
+        return result, embeds
+
+    return result
+
+
+def stage1_loss(
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:121]: Can we mention here that stage 1 training is much lighter weight than stage 2 training, so users know it will take much longer in the second stage

File: speculator/train_speculator_utils.py [hunk 21]
```diff
@@ -0,0 +1,546 @@
...
+    Given the sequence of embeddings produced in parallel by the base model,
+    get n+2,n+3,... speculator predictions and compare to ground truth tokens.
+    ...
+    Args
+    ----
+    model: nn.Module
+        The frozen base model. Must return output logits AND corresponding embedding vectors.
+    speculator: nn.Module
+        The speculator to be trained. Takes as input sequence of embeddings and token indices,
+        and return token prediction logits for each head.
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:135]: missing base_model_input in docs

File: speculator/train_speculator_utils.py [hunk 22]
```diff
@@ -0,0 +1,546 @@
...
+    input: torch.IntTensor
+        The ground truth token indices.
+    loss_fn: Callable
+        Torch loss function comparing logits to indices i.e. CrossEntropyLoss()
+    ddp_stats: torch.FloatTensor
+        Aggregate stat tracking buffer.
+        Entries are: grad norm, accumulation steps, head 1 loss, head 2 loss, etc.
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:142]: missing base_model_mesh in docs

File: speculator/train_speculator_utils.py [hunk 23]
```diff
@@ -0,0 +1,546 @@
...
+):
+    """
+    Perform a forward pass for stage 2 training and calculate the loss.
+    Given the sequence of embeddings produced in serial by the base model,
+    get n+1,n+2,... speculator predictions and compare to base model's generated tokens.
+    Reshapes input to more entries / shorter sequences, for more efficient generation.
+    ...
+    Args
+    ----
+    cfg: train_config
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:179]: add missing arguments here for docs

File: speculator/train_speculator_utils.py [hunk 24]
```diff
@@ -0,0 +1,546 @@
...
+        batch_idx,
+        speculator,
+        tokens_seen=elapsed_tokens + n_tok,
+        is_compiled=cfg.use_torch_compile,
+    )
+
+    return train_loss
+
+
+class EmbedGPTBigCode(GPTBigCode):
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:415]: may want to remove these in favor of a HiddenStatesExtractor wrapper, but not crucial for now. This way we can remove all the extra variants and adapters.

File: fms_fsdp/config/training.py (hunk 1 above)
Comment (sahilsuneja1) [reply fms_fsdp/config/training.py:62]: Separate config not needed. Many of the above are bring used, and the remaining ones are also compatible.

File: fms_fsdp/utils/dataloader_utils.py (hunk 6 above)
Comment (sahilsuneja1) [reply fms_fsdp/utils/dataloader_utils.py:71]: Sorry, I didn't follow

File: speculator/train_speculator.py (hunk 8 above)
Comment (sahilsuneja1) [reply speculator/train_speculator.py:34]: What would you check in the assert?

File: speculator/train_speculator.py (hunk 9 above)
Comment (sahilsuneja1) [reply speculator/train_speculator.py:115]: Will revisit this

File: speculator/train_speculator.py (hunk 10 above)
Comment (sahilsuneja1) [reply speculator/train_speculator.py:117]: Will include that in next iteration after testing

File: speculator/train_speculator.py (hunk 15 above)
Comment (sahilsuneja1) [reply speculator/train_speculator.py:259]: @daviswer?

File: speculator/train_speculator_utils.py (hunk 21 above)
Comment (sahilsuneja1) [reply speculator/train_speculator_utils.py:135]: Will add

File: fms_fsdp/utils/dataloader_utils.py (hunk 5 above)
Comment (nairbv) [reply fms_fsdp/utils/dataloader_utils.py:27]: nit: I think it's generally considered preferable "pytorch style" to use torch.tensor(..., dtype=torch.int) over IntTensor

File: fms_fsdp/utils/dataloader_utils.py (hunk 6 above)
Comment (nairbv) [reply fms_fsdp/utils/dataloader_utils.py:71]: parameterize the callable (`List[Callable[[Parameter1Type, Parameter2Type], ReturnType]]`) for whatever the parameters/return are

Comment (sahilsuneja1) [reply fms_fsdp/utils/dataloader_utils.py:71]: @daviswer

File: fms_fsdp/utils/dataloader_utils.py (hunk 5 above)
Comment (sahilsuneja1) [reply fms_fsdp/utils/dataloader_utils.py:27]: Will do!

Comment (daviswer) [reply fms_fsdp/utils/dataloader_utils.py:27]: This doesn't need device - it's internal to the `torch.utils.data.Dataset` pipeline, and all pipeline stuff is on cpu until the data leaves the pipeline (that's a general pytorch rule, not just us)

File: speculator/train_speculator.py (hunk 8 above)
Comment (JRosenkranz) [reply speculator/train_speculator.py:34]: You could have a sample input/output file that the output should match the expected output

File: speculator/train_speculator.py (hunk 11 above)
Comment (JRosenkranz) [reply speculator/train_speculator.py:136]: Maybe we could have a test json that can be used (by default not required) and the json would include prompt, expected output, etc. Then we could just add to the config test_metadata param

File: fms_fsdp/utils/dataloader_utils.py (hunk 6 above)
Comment (JRosenkranz) [reply fms_fsdp/utils/dataloader_utils.py:71]: Maybe an example would help if that can be added here in the docs

File: speculator/train_speculator.py (hunk 15 above)
Comment (daviswer) [reply speculator/train_speculator.py:260]: Added some comments

File: speculator/train_speculator_utils.py (hunk 19 above)
Comment (JRosenkranz) [reply speculator/train_speculator_utils.py:28]: Ah, nvm this would be good to figure out, but as it stands currently we either return logits and past_key_values, there is no hook to get the embeds and include them.

File: speculator/train_speculator.py [hunk 25]
```diff
@@ -0,0 +1,329 @@
...
+    # some setups
+    torch.cuda.set_device(local_rank)
+
+    if cfg.sharding_strategy != "tp":
+        setup()
+        torch._C._distributed_c10d._register_process_group("default", dist.group.WORLD)
+        base_model_mesh = None
+        speculator_mesh = None
+    else:
+        tp_size = 8
```
[the remark this answers is not in the document]
Comment (sahilsuneja1) [reply speculator/train_speculator.py]: You are right-- added as config param

**Comment by sahilsuneja1** [comment]
Closing in favor of #114
